packages/flutter/lib/src/widgets/basic.dart

```diff
@@ -2315,6 +2316,201 @@ class ConstrainedBox extends SingleChildRenderObjectWidget {
+/// This container sizes its child using a [BoxConstraints] created by applying
+/// [constraintsTransform] to its own constraints. This container will then
+/// attempt to adopt the same size, within the limits of its own constraints. If
+/// it ends up with a different size, it will align the child based on
+/// [alignment]. If the container cannot expand enough to accommodate the entire
+/// child, the child will be clipped if [clipBehavior] is not [Clip.none].
+///
+/// In debug mode, if the child overflows the container, a warning will be
+/// printed on the console, and black and yellow striped areas will appear where
+/// the overflow occurs.
+///
+/// {@tool snippet}
+/// In the following snippet, the child [Card] is guaranteed to be at least as
+/// tall as its intrinsic height. Unlike [UnconstrainedBox], the child can be
+/// taller if the the parent's minHeight is higher than the child's intrinsic
+/// height. If parent container's maxHeight is less than [Card]'s intrinsic
+/// height, in debug mode a warning will be given.
+///
+/// ```dart
+/// ConstraintsTransformBox(
+///   constraintsTransform: (BoxConstraints constraints) => constraints.copyWith(maxHeight: double.infinity),
+///   child: const Card(child: Text('Hello World!')),
+/// )
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [ConstrainedBox], which renders a box which imposes constraints
+///    on its child.
+///  * [OverflowBox], a widget that imposes additional constraints on its child,
+///    and allows the child to overflow itself.
+///  * [UnconstrainedBox] which allows its children to render themselves
+///    unconstrained, expands to fit them, and considers overflow to be an error.
+class ConstraintsTransformBox extends SingleChildRenderObjectWidget {
+  /// Creates a widget that uses a function to transform the constraints it
+  /// passes to its child. If the child overflows the parent's constraints, a
+  /// warning will be given in debug mode.
+  ///
+  /// The [debugTransformType] argument adds a debug label to this widget.
+  ///
+  /// The `alignment`, `clipBehavior` and `constraintsTransform` arguments must
+  /// not be null.
+  const ConstraintsTransformBox({
+    Key? key,
+    Widget? child,
+    this.textDirection,
+    this.alignment = Alignment.center,
+    required this.constraintsTransform,
+    this.clipBehavior = Clip.none,
+    String debugTransformType = '',
+  }) : _debugTransformLabel = debugTransformType,
+       assert(alignment != null),
+       assert(clipBehavior != null),
+       assert(constraintsTransform != null),
+       super(key: key, child: child);
+
+  /// Creates a widget that imposes no constraints on its child, allowing it to
+  /// render at its "natural" size. If the child overflows the parent's
+  /// constraints, a warning will be given in debug mode.
+  ///
+  /// The `alignment`, `clipBehavior` and `constraintsTransform` arguments must
+  /// not be null.
+  const ConstraintsTransformBox.unconstrained({
```

Comment:
Isn't this just duplicating behavior of the UnconstrainedBox widget? Can we just link to that one in the docs?

Reply:
Same with the other two here.

Reply:
I'll expose the static functions instead of named constructors.